cmake/CMakeLists.txt

```diff
@@ -231,10 +231,14 @@ if (NOT (onnxruntime_ENABLE_PYTHON AND onnxruntime_ENABLE_TRAINING))
   set(onnxruntime_ENABLE_TRAINING_TORCH_INTEROP OFF)
 endif()
 
+set(onnxruntime_REQUIRE_PYTHON_EMBED_LIB OFF)
 if (onnxruntime_ENABLE_TRAINING_TORCH_INTEROP)
   add_compile_definitions(ENABLE_TRAINING_TORCH_INTEROP)
-  # disable tests because executable projects need link to python so which did not always exist for test envs.
-  set(onnxruntime_BUILD_UNIT_TESTS OFF)
+
+  # Python::Python is required for building unit test executables or onnxruntime.so. 
+  if (onnxruntime_BUILD_UNIT_TESTS OR onnxruntime_BUILD_SHARED_LIB)
+    set(onnxruntime_REQUIRE_PYTHON_EMBED_LIB ON)
```

Comment:
Why onnxruntime_BUILD_SHARED_LIB needs this? 

Whenever onnxruntime_REQUIRE_PYTHON_EMBED_LIB is ON, mac build is broken. Because we can't distribute the python lib which came from macOS.

Reply:
I saw some unresolved symbols when compiling onnxruntime.cmake using Python::Module. Do you think that is expected? So I add a check for onnxruntime_BUILD_SHARED_LIB

Reply:
Let's try to get more details.

Reply:
I reproed the issue. 

[ 98%] Built target onnxruntime_providers
[ 98%] Built target onnxruntime_providers_cuda
Consolidate compiler generated dependencies of target onnxruntime
Consolidate compiler generated dependencies of target onnxruntime_pybind11_state
[ 98%] Linking CXX shared module libonnxruntime.so
[100%] Built target onnxruntime_pybind11_state
libonnxruntime_providers.a(torch_custom_function_kernel_base.cc.o): In function `onnxruntime::contrib::PythonOpBase::AddIntScalarArgs()':
/home/pengwa/dev/onnxruntime/orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc:120: undefined reference to `Py_BuildValue'
/home/pengwa/dev/onnxruntime/orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc:125: undefined reference to `Py_BuildValue'
libonnxruntime_providers.a(torch_custom_function_kernel_base.cc.o): In function `onnxruntime::contrib::PythonOpBase::AddInputTupleArgs()':
/home/pengwa/dev/onnxruntime/orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc:137: undefined reference to `PyTuple_New'
/home/pengwa/dev/onnxruntime/orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc:139: undefined reference to `Py_BuildValue'
/home/pengwa/dev/onnxruntime/orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc:140: undefined reference to `PyTuple_SetItem'
libonnxruntime_providers.a(torch_custom_function_kernel_base.cc.o): In function `onnxruntime::contrib::PythonOpBase::AddFloatTupleArgs()':
/home/pengwa/dev/onnxruntime/orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc:155: undefined reference to `PyTuple_New'
/home/pengwa/dev/onnxruntime/orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc:157: undefined reference to `Py_BuildValue'
/home/pengwa/dev/onnxruntime/orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc:158: undefined reference to `PyTuple_SetItem'
libonnxruntime_interop_torch.a(torch_proxy.cc.o): In function `onnxruntime::language_interop_ops::torch::InvokeRunner(_object*, _object*, bool, void**, std::vector<OrtValue, std::allocator<OrtValue> >&)':
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:138: undefined reference to `PyObject_CallObject'
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:140: undefined reference to `PyErr_Occurred'
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:151: undefined reference to `PyTuple_GetItem'
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:164: undefined reference to `_Py_NoneStruct'
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:172: undefined reference to `PyTuple_Size'
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:173: undefined reference to `PyTuple_GetItem'
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:141: undefined reference to `PyErr_Print'
libonnxruntime_interop_torch.a(torch_proxy.cc.o): In function `onnxruntime::language_interop_ops::torch::Ort_PyTuple_New(unsigned long, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:20: undefined reference to `PyTuple_New'
libonnxruntime_interop_torch.a(torch_proxy.cc.o): In function `onnxruntime::language_interop_ops::torch::Ort_PyList_New(unsigned long, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:37: undefined reference to `PyList_New'
libonnxruntime_interop_torch.a(torch_proxy.cc.o): In function `onnxruntime::language_interop_ops::torch::Ort_PyTuple_SetItem_NoIncref(_object*, unsigned long, _object*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:33: undefined reference to `PyTuple_SetItem'
libonnxruntime_interop_torch.a(torch_proxy.cc.o): In function `onnxruntime::language_interop_ops::torch::Ort_PyList_SetItem_NoIncref(_object*, unsigned long, _object*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:50: undefined reference to `PyList_SetItem'
libonnxruntime_interop_torch.a(torch_proxy.cc.o): In function `onnxruntime::language_interop_ops::torch::CreateTensorFlags(unsigned long, std::vector<long, std::allocator<long> > const&)':
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:104: undefined reference to `PyLong_FromLong'
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:109: undefined reference to `PyLong_FromLong'
libonnxruntime_interop_torch.a(torch_proxy.cc.o): In function `onnxruntime::language_interop_ops::torch::Ort_PyList_SetItem_Incref(_object*, unsigned long, _object*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:45: undefined reference to `PyList_SetItem'
libonnxruntime_interop_torch.a(torch_proxy.cc.o): In function `onnxruntime::language_interop_ops::torch::CreateRequiresGradFlags(std::vector<long, std::allocator<long> > const&)':
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:125: undefined reference to `_Py_FalseStruct'
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:123: undefined reference to `_Py_TrueStruct'
libonnxruntime_interop_torch.a(torch_proxy.cc.o): In function `onnxruntime::language_interop_ops::torch::Ort_PyTuple_SetItem_Incref(_object*, unsigned long, _object*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:28: undefined reference to `PyTuple_SetItem'
libonnxruntime_interop_torch.a(torch_proxy.cc.o): In function `onnxruntime::language_interop_ops::torch::CreatePythonCallArguments(_object*, unsigned long, std::vector<long, std::allocator<long> > const&, std::vector<OrtValue, std::allocator<OrtValue> > const&, std::vector<long, std::allocator<long> > const&, std::vector<void*, std::allocator<void*> > const&, std::vector<long, std::allocator<long> > const&, bool, bool)':
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:191: undefined reference to `PyCallable_Check'
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:206: undefined reference to `_Py_TrueStruct'
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:208: undefined reference to `_Py_TrueStruct'
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:208: undefined reference to `_Py_FalseStruct'
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:206: undefined reference to `_Py_FalseStruct'
libonnxruntime_interop_torch.a(torch_proxy.cc.o): In function `onnxruntime::language_interop_ops::torch::TorchProxy::Forward(void*, std::vector<long, std::allocator<long> > const&, std::vector<OrtValue, std::allocator<OrtValue> > const&, std::vector<long, std::allocator<long> > const&, std::vector<void*, std::allocator<void*> > const&, std::vector<long, std::allocator<long> > const&, void**, std::vector<OrtValue, std::allocator<OrtValue> >&, bool, bool)':
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/gil.h:16: undefined reference to `PyGILState_Ensure'
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/gil.h:17: undefined reference to `PyGILState_Release'
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/gil.h:17: undefined reference to `PyGILState_Release'
libonnxruntime_interop_torch.a(torch_proxy.cc.o): In function `onnxruntime::language_interop_ops::torch::TorchProxy::Backward(void*, std::vector<long, std::allocator<long> > const&, std::vector<OrtValue, std::allocator<OrtValue> > const&, std::vector<long, std::allocator<long> > const&, std::vector<void*, std::allocator<void*> > const&, std::vector<long, std::allocator<long> > const&, std::vector<OrtValue, std::allocator<OrtValue> >&, bool)':
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/gil.h:16: undefined reference to `PyGILState_Ensure'
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/gil.h:17: undefined reference to `PyGILState_Release'
/home/pengwa/dev/onnxruntime/onnxruntime/core/language_interop_ops/torch/gil.h:17: undefined reference to `PyGILState_Release'
libonnxruntime_python_interface.a(dlpack_python.cc.o): In function `onnxruntime::dlpack::DlpackCapsuleDestructor(_object*)':
/home/pengwa/dev/onnxruntime/onnxruntime/core/dlpack/dlpack_python.cc:11: undefined reference to `PyCapsule_GetPointer'
/home/pengwa/dev/onnxruntime/onnxruntime/core/dlpack/dlpack_python.cc:18: undefined reference to `PyErr_Clear'
libonnxruntime_python_interface.a(dlpack_python.cc.o): In function `onnxruntime::dlpack::ToDlpack(OrtValue)':
/home/pengwa/dev/onnxruntime/onnxruntime/core/dlpack/dlpack_python.cc:24: undefined reference to `PyCapsule_New'
libonnxruntime_python_interface.a(dlpack_python.cc.o): In function `onnxruntime::dlpack::FromDlpack(_object*, bool)':
/home/pengwa/dev/onnxruntime/onnxruntime/core/dlpack/dlpack_python.cc:29: undefined reference to `PyCapsule_GetPointer'
/home/pengwa/dev/onnxruntime/onnxruntime/core/dlpack/dlpack_python.cc:32: undefined reference to `PyCapsule_SetName'
collect2: error: ld returned 1 exit status
CMakeFiles/onnxruntime.dir/build.make:128: recipe for target 'libonnxruntime.so' failed
make[2]: *** [libonnxruntime.so] Error 1
CMakeFiles/Makefile2:1533: recipe for target 'CMakeFiles/onnxruntime.dir/all' failed
make[1]: *** [CMakeFiles/onnxruntime.dir/all] Error 2
Makefile:165: recipe for target 'all' failed
make: *** [all] Error 2

Reply:
after investigation, it it is found in onnxruntime.cmake, there is a --no-undefined link options. So that's reason we see the same link approach. onxnruntiome_python.cmake works, but onnxruntime.cmake did not. So after removing it, the build pass. So do you think it is okay to remove this option as below? 

![image](https://user-images.githubusercontent.com/10530022/121999011-4e4cf700-cddf-11eb-996a-ee856c371bed.png)

Reply:
> Why onnxruntime_BUILD_SHARED_LIB needs this?
> 
> Whenever onnxruntime_REQUIRE_PYTHON_EMBED_LIB is ON, mac build is broken. Because we can't distribute the python lib which came from macOS.

I think we don;t need let any Mac build to enable the torch interop. So it should not be a problem. Thanks for the reminder.